cpp/ql/src/experimental/Security/CWE/CWE-476/DangerousUseOfExceptionBlocks.ql

```diff
@@ -0,0 +1,165 @@
+/**
+ * @name Dangerous use of exception blocks.
+ * @description When clearing the data in the catch block, you must be sure that the memory was allocated before the exception.
+ * @kind problem
+ * @id cpp/dangerous-use-of-exception-blocks
+ * @problem.severity warning
+ * @precision medium
+ * @tags correctness
+ *       security
+ *       external/cwe/cwe-476
+ *       external/cwe/cwe-415
+ */
+
+import cpp
+
+/** Holds if the release can occur twice. in the current block of catch and above in the block of try or other block catch. */
+pragma[inline]
+predicate doubleCallDelete(CatchAnyBlock cb, Variable vr) {
+  // Search for exceptions after freeing memory.
+  exists(Expr e1 |
+    (
+      e1 = vr.getAnAccess().getEnclosingStmt().(ExprStmt).getExpr().(DeleteArrayExpr) or
+      e1 = vr.getAnAccess().getEnclosingStmt().(ExprStmt).getExpr().(DeleteExpr)
+    ) and
+    e1.getEnclosingFunction() = cb.getEnclosingFunction() and
+    (
+      e1.getEnclosingStmt().getParentStmt*() = cb.getTryStmt().getStmt() and
+      exists(Expr e2, ThrowExpr th |
+        (
+          e2 = th or
+          e2 = th.getEnclosingFunction().getACallToThisFunction()
+        ) and
+        e2.getEnclosingStmt().getParentStmt*() = cb.getTryStmt().getStmt() and
+        e1.getASuccessor+() = e2
+      ) and
+      not exists(AssignExpr ae |
+        ae.getLValue().(VariableAccess).getTarget() = vr and
+        ae.getRValue().getValue() = "0" and
+        e1.getASuccessor+() = ae and
+        ae.getEnclosingStmt().getParentStmt*() = cb.getTryStmt().getStmt()
+      )
+      or
+      // Search for a situation when there is a higher catch block that also frees memory.
+      exists(CatchBlock cbt, Expr e2, ThrowExpr th |
+        e1.getEnclosingStmt().getParentStmt*() = cbt and
+        exists(cbt.getParameter()) and
+        (
+          e2 = th or
+          e2 = th.getEnclosingFunction().getACallToThisFunction()
+        ) and
+        e2.getEnclosingStmt().getParentStmt*() = cbt and
+        e1.getASuccessor+() = e2 and
+        not exists(AssignExpr ae |
+          ae.getLValue().(VariableAccess).getTarget() = vr and
+          ae.getRValue().getValue() = "0" and
+          e1.getASuccessor+() = ae and
+          ae.getEnclosingStmt().getParentStmt*() = cbt
+        )
+      )
+    ) and
+    // Exclude the presence of a check in catch block.
+    not exists(IfStmt ifst |
+      ifst.getEnclosingStmt().getParentStmt*() = cb.getAStmt()
+    )
+  )
+}
```

Comment:
I think there are several paths to simplifying this function and reducing the amount of duplicate code in it.  One would be to add an additional parameter to the predicate, for the block `b` where `vr` is deleted:
```
/** Holds if `vr` may be released in the block `b`, which is the `try` block associated with `cb` or a `catch` block prior to `cb`. */
predicate doubleCallDelete(Block b, CatchAnyBlock cb, Variable vr) {
```
In the first case we can set it with:
```
b = cb.getTryStmt().getStmt()
````
in the second it replaces the variable `cbt`, we just have to cast it to `CatchBlock` in the appropriate place:
```
exists(b.(CatchBlock).getParameter()) and
```
Having done this, both copies of the code for spotting assignments `vr = 0` can be combined into one.  I think the logic around `e2` can also be combined into one copy as well.

Reply:
thanks for the idea.
I hope I did as you intended?

Reply:
Yep, the change looks good to me!